include/nbl/builtin/hlsl/spirv_intrinsics/raytracing.hlsl

```diff
@@ -0,0 +1,48 @@
+// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_SPIRV_INTRINSICS_RAYTRACING_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SPIRV_INTRINSICS_RAYTRACING_INCLUDED_
+
+// #include "nbl/builtin/hlsl/spirv_intrinsics/core.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace spirv
+{
+
+[[vk::ext_capability(spv::CapabilityRayQueryKHR)]]
+[[vk::ext_instruction(spv::OpRayQueryInitializeKHR)]]
+void rayQueryInitializeEXT(/* spv::OpTypeRayQueryKHR function_ptr query, spv::OpTypeAccelerationStructureKHR AS, uint32_t flags, uint32_t cull mask, float3 origin, float32_t tmin, float3 direction, float32_t tmax*/);
+
+[[vk::ext_capability(spv::CapabilityRayQueryKHR)]]
+[[vk::ext_instruction(spv::OpRayQueryProceedKHR)]]
+bool rayQueryProceedEXT(/* spv::OpTypeRayQueryKHR function_ptr query */);
+
+[[vk::ext_capability(spv::CapabilityRayQueryKHR)]]
+[[vk::ext_instruction(spv::OpRayQueryGetIntersectionTypeKHR)]]
+int rayQueryGetIntersectionTypeEXT(/* spv::OpTypeRayQueryKHR function_ptr query, uint32_t intersection (candidate 0, commited 1) */);
+
+[[vk::ext_capability(spv::CapabilityRayQueryKHR)]]
+[[vk::ext_instruction(spv::OpRayQueryGetIntersectionInstanceIdKHR)]]
+int rayQueryGetIntersectionInstanceIdEXT(/* spv::OpTypeRayQueryKHR function_ptr query, uint32_t intersection (candidate 0, commited 1) */);
+
+[[vk::ext_capability(spv::CapabilityRayQueryKHR)]]
+[[vk::ext_instruction(spv::OpRayQueryGetIntersectionPrimitiveIndexKHR)]]
+int rayQueryGetIntersectionPrimitiveIndexEXT(/* spv::OpTypeRayQueryKHR function_ptr query, uint32_t intersection (candidate 0, commited 1) */);
```

Comment:
I think `spv::OpTypeRayQueryKHR function_ptr query` should be expressible with `pointer_t` https://github.com/Devsh-Graphics-Programming/Nabla/blob/38491b8407dc1ff49d133be319f599cf60e59342/include/nbl/builtin/hlsl/spirv_intrinsics/core.hlsl#L53

Reply:
alternatively you can just try `[vk::ext_reference] T`

Reply:
also shouldn't the suffices on these functions be `KHR` not `EXT` ?

Reply:
for the acceeleration structures use the `ext_reference` of the HLSL Acceleration Structure type because of https://github.com/microsoft/DirectXShaderCompiler/issues/7014

Reply:
I had an idea to clean this up and not involve HLSL types in our `spirv` namespace and do this cleanly
```cpp
namespace spirv
{
// unholy disgusting thing
template<uint32_t StorageClass, typename T, typename U>
[[vk::ext_instruction(spv::OpCopyObject)]]
pointer_t<StorageClass,T> copyObject([[vk::ext_reference]] U v);

//[[vk::ext_capability(spv::CapabilityRayQueryKHR)]]
using RayQueryKHR = vk::SpirvOpaqueType<spv::OpTypeRayQueryKHR>;

//[[vk::ext_capability(spv::CapabilityAccelerationStructureKHR)]]
using AccelerationStructureKHR = vk::SpirvOpaqueType<spv::OpTypeAccelerationStructureKHR>;


[[vk::ext_capability(spv::CapabilityRayQueryKHR)]]
[[vk::ext_instruction(spv::OpRayQueryInitializeKHR)]]
void rayQueryInitializeKHR([[vk::ext_reference]] RayQueryKHR query, [[vk::ext_reference]] AccelerationStructureKHR AS, uint32_t flags, uint32_t cull_mask, float32_t3 origin, float32_t tmin, float32_t3 direction, float32_t tmax);
}

namespace glsl
{
template<uint32_t FLAGS>
void rayQueryInitializeKHR(NBL_REF_ARG(RayQuery<FLAGS>) query, NBL_CONST_REF_ARG(RaytracingAccelerationStructure) AS, const uint32_t cull_mask, const float32_t3 origin, const float32_t tmin, const float32_t3 direction, const float32_t tmax)
{
    spirv::rayQueryInitializeKHR(spirv::copyObject<spv::StorageClassFunction,spirv::RayQueryKHR>(query),spirv::copyObject<spv::StorageClassFunction,spirv::RayQueryKHR>(AS),FLAGS,cull_mask,origin,tmin,direction,tmax);
}
}
```
bridging the HLSL with spirv via the GLSL namespace, and properly reinterepret_cast-ing the HLSL types into their SPIR-V variable poitners in correct storage spaces, but alas another DXC limitation https://github.com/microsoft/DirectXShaderCompiler/issues/7015

So seems for every intrinsic argument that is a `AccelerationStructureKHR ` you need two overloads one for `RaytracingAccelerationStructure` and another for `AccelerationStructureKHR`

Reply:
> also shouldn't the suffices on these functions be `KHR` not `EXT` ?

I named them following the [GLSL equivalents](https://github.com/KhronosGroup/GLSL/blob/main/extensions/ext/GLSL_EXT_ray_query.txt). Could change it to `KHR`

Reply:
don't look at GLSL, only look at SPIR-V, we have a separate namespace for GLSL stuff (which we probably won't touch right now).